File: 1588. Sum of All Odd Length Subarrays/solution.js
var sumOddLengthSubarrays = function (arr) {
	// first all single digit subarrays
	let total = 0;
	// go through array
	for (let i = 0; i < arr.length; i++) {
		//generate all sub arrays
		for (let j = i; j < arr.length; j += 2) {
			for (let k = i; k <= j; k++) total += arr[k];
		}
	}
  return total; 

	/* An alternative method is to figure out how many times each number will
appear in all the various arrays, and then mutltiply the number by that frequency.

Outer array numbers appear less freqently than inner numbers, in the amount of 
1/2 of N for even-N arrays and 1/2 + .5 for odd-N arrays, and so on for the others
So a frequency equation for an array of N length is this:
Math.floor((((i + 1) * (N - i))+ 1) / 2);

But honestly, the math on this is SOOO confusing, the nested for loops is better for humans ... and a bit faster?

	let N = arr.length;
	for (let i = 0; i < N; i++) {
		total += arr[i] * Math.floor(((i + 1) * (N - i) + 1) / 2);
	}

	return total;
  */
};

const arr = [1, 2, 3, 4];
console.log(sumOddLengthSubarrays(arr));

/* Given an array of positive integers arr, return the sum of all possible
odd-length subarrays of arr.
[1,4,2,5,3] = 58
[1,2] = 3
[1,2,3] = 12
[1,2,3,4] = 10 + 6 + 9 = 25
[10,11,12] = 66
*/

File: 1588. Sum of All Odd Length Subarrays/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

int sumOddLengthSubarrays(vector<int> &arr) {
  int arrLen = arr.size();
  int arrMid = (arr.size() / 2);
  int arrSum = 0;

  if (arrLen <= 2) {
    for (int i = 0; i < arrLen; i++)
      arrSum += arr[i];
    return arrSum;
  } else {
    for (int i = 0; i < arrLen / 2; i++) {
      arrSum = arrSum + (arr[i] * (arrLen + i - 2)) +
               (arr[arrLen - i] * (arrLen + i - 2));
    }
  }
  return arrSum;
}

/*
 else if (aLen % 2 == 0) {
  for (int i = 0; i < aLen; i++)
    arrSum += arr[i];
}
 else {
 for (int i = 0; i < aLen; i++) {
   arrSum += arr[i];
 }
 arrSum *= 2;
}
  */

int main() {
  vector<int> arr = {1, 4, 2, 5, 3};
  cout << sumOddLengthSubarrays(arr);
}

/* Given an array of positive integers arr, return the sum of all possible
odd-length subarrays of arr.
[1,4,2,5,3] = 58
[1,2] = 3
[1,2,3] = 12
[1,2,3,4] = 10 + 6 + 9 = 25
[10,11,12] = 66
*/
